### PR TITLE
added red warning text

### PR DESCRIPTION
**Pull request by hardikparashar**
### What do you want to achieve?
-> Display text more than 256 characters long with a red background and disable the send message functionality for the user while chatting with a dialogflow bot.

### How was the code tested?

->

-----------
Screenshot 2019-09-09 at 8 20 25 PM
------------

### What new thing you came across while writing this code? 
-> Selection and range in javascript.

### Incase you fixed a bug then please describe the root cause of it? 
-> no bugs fixed.

NOTE: Make sure you're comparing your branch with the correct base branch

### CODE REVIEW

File: webplugin/js/app/mck-sidebox-1.0.js [hunk 1]
```diff
@@ -2480,19 +2480,88 @@ var CURRENT_GROUP_DATA={};
...
-                        var str = ($mck_text_box.text().toString()).trim();
-                        var len = str.length;
-                        if (len > 199) {
-                            if (len > 256) {
-                                remtxt = "(" + MCK_LABELS['limit.remove'] + " " + (len - 256) + " " + MCK_LABELS['limit.characters'] + ")";
+                        let textVal = mckUtils.textVal(textBox);
+                        let str = textVal.trim();
+                        var textLength = str.length;
+                        if (textLength > warningLength) {
+                            let caretPosition = _this.cursorPosition();
```
Comment (devashishmamgain) [inline webplugin/js/app/mck-sidebox-1.0.js]: @hardikparashar for web plugin, please use 'var'. let might not work in IE.

File: webplugin/js/app/mck-sidebox-1.0.js [hunk 2]
```diff
@@ -2480,19 +2480,88 @@ var CURRENT_GROUP_DATA={};
                     _this.toggleMediaOptions(this);
 
                 });
-                $mck_text_box.keyup(function (){
+             
+            _this.cursorPosition =  function () {
+                var sel = document.getSelection();
+                sel.modify("extend", "backward", "documentboundary");
```
Comment (CaptainT33mo) [inline webplugin/js/app/mck-sidebox-1.0.js]: @hardikparashar `.modify()` will not work in Internet Explorer and MS Edge. Need to find some other solution. https://developer.mozilla.org/en-US/docs/Web/API/Selection/modify

File: webplugin/js/app/mck-sidebox-1.0.js [hunk 3]
```diff
@@ -2480,19 +2480,88 @@ var CURRENT_GROUP_DATA={};
...
+                var pos = sel.toString().length;
+                if(sel.anchorNode != undefined) sel.collapseToEnd();
```
Comment (ab14bhardwaj) [inline webplugin/js/app/mck-sidebox-1.0.js]: @hardikparashar collapseToEnd doesn't work in IE.
https://developer.mozilla.org/en-US/docs/Web/API/Selection/collapseToEnd

Comment (akshatdhabi) [inline webplugin/js/app/mck-sidebox-1.0.js]: .anchorNode is not supported in IE https://developer.mozilla.org/en-US/docs/Web/API/Selection/anchorNode

File: webplugin/js/app/mck-sidebox-1.0.js [hunk 4]
```diff
@@ -2480,19 +2480,88 @@ var CURRENT_GROUP_DATA={};
...
+                let sendButton = document.getElementById('mck-msg-sbmt');
+                sendButton.setAttribute("disabled", true);
+                CURRENT_GROUP_DATA.DISABLE_SEND_MESSAGE = true;
+            }
+            _this.enableSendButton = function () {
+                let sendButton = document.getElementById('mck-msg-sbmt');
+                sendButton.removeAttribute("disabled", false);
+                CURRENT_GROUP_DATA.DISABLE_SEND_MESSAGE = false;
+            }
+                $mck_text_box.on('input paste', function (event){
```
Comment (akshatdhabi) [inline webplugin/js/app/mck-sidebox-1.0.js]: please refrain from using jquery

File: webplugin/js/app/mck-sidebox-1.0.js (hunk 3 above)
Comment (ab14bhardwaj) [reply webplugin/js/app/mck-sidebox-1.0.js]: It works, just checked it on below link. Resolving the comment.
https://msdn.microsoft.com/en-us/ie/ff975174(v=vs.94)

File: webplugin/js/app/mck-sidebox-1.0.js (hunk 1 above)
Comment (devashishmamgain) [reply webplugin/js/app/mck-sidebox-1.0.js]: @hardikparashar there are more places where let is used. please remove from all places.

File: webplugin/js/app/mck-sidebox-1.0.js [hunk 5]
```diff
@@ -2480,19 +2480,101 @@ var CURRENT_GROUP_DATA={};
...
+                    var position = selection.toString().length;
+                    selection.collapseToEnd();
+                    return position;
+                }
+                _this.disableSendButton = function () {
+                    var sendButton = document.getElementById('mck-msg-sbmt');
+                    sendButton.setAttribute("disabled", true);
+                    CURRENT_GROUP_DATA.DISABLE_SEND_MESSAGE = true;
+                }
+                _this.enableSendButton = function () {
```
Comment (devashishmamgain) [inline webplugin/js/app/mck-sidebox-1.0.js]: @hardikparashar combine disableSendButton and enableSendButton into a single function, its the same code with just a boolean difference

File: webplugin/js/app/mck-sidebox-1.0.js [hunk 6]
```diff
@@ -2480,19 +2480,101 @@ var CURRENT_GROUP_DATA={};
...
+                                var normalSpan = document.getElementById('mck-normal-span');
+                                if (warningSpan && !(event.type == 'paste')) {
+                                    if(caretPosition <= (maxLength + 1)) {
+                                        warningSpan.innerHTML = spanContent;
+                                        normalSpan.innerHTML = preSpanContent;
+                                        selection = document.getSelection();
+                                        (caretPosition > maxLength) ? selection.collapse(warningSpan.firstChild,caretPosition - maxLength) : selection.collapse(normalSpan.firstChild,caretPosition);
+                                    }
+                                } else if(warningSpan && (event.type == 'paste')) {
+                                    textBox.innerHTML = '<span id = "mck-normal-span"></span><span contenteditable="true" style="background-color:rgb(255,184,194);" id="mck-text-warning-span"></span>';
```
Comment (devashishmamgain) [inline webplugin/js/app/mck-sidebox-1.0.js]: @hardikparashar can this "background-color:rgb(255,184,194);" be moved to a css file?

File: webplugin/js/app/mck-sidebox-1.0.js [hunk 7]
```diff
@@ -2480,19 +2480,96 @@ var CURRENT_GROUP_DATA={};
...
+                                if (warningSpan && !(event.type == 'paste')) {
+                                    if(caretPosition <= (maxLength + 1)) {
+                                        warningSpan.innerHTML = spanContent;
+                                        normalSpan.innerHTML = preSpanContent;
+                                        selection = document.getSelection();
+                                        (caretPosition > maxLength) ? selection.collapse(warningSpan.firstChild,caretPosition - maxLength) : selection.collapse(normalSpan.firstChild,caretPosition);
+                                    }
+                                } else if(warningSpan && (event.type == 'paste')) {
+                                    textBox.innerHTML = '<span id = "mck-normal-span"></span><span contenteditable="true"  id="mck-text-warning-span"></span>';
+                                    normalSpan = document.getElementById('mck-normal-span');
```
[the remark this answers is not in the document]
Comment (hardikparashar) [reply webplugin/js/app/mck-sidebox-1.0.js]: It's a dynamic component, everytime it gets generated need to re-select it

File: webplugin/js/app/mck-sidebox-1.0.js [hunk 8]
```diff
@@ -2480,19 +2480,135 @@ var CURRENT_GROUP_DATA={};
                     _this.toggleMediaOptions(this);
 
                 });
-                $mck_text_box.keyup(function (){
+                 _this.cursorPosition = function (element) {
+                    var selection = window.getSelection();
+                    var nodeCount = 0;
+                    var emptyNodes=0;
```
Comment (devashishmamgain) [inline webplugin/js/app/mck-sidebox-1.0.js]: @hardikparashar please format the code

File: webplugin/js/app/mck-sidebox-1.0.js [hunk 9]
```diff
@@ -2480,19 +2480,135 @@ var CURRENT_GROUP_DATA={};
...
+                    var currentNode = selection.focusNode;
+                    if(normalSpan) {
+                        startNode = normalSpan.firstChild;
+                    }
+                    var newRange = document.createRange();
+                    newRange.setStart(startNode,0);
+                    newRange.setEnd(currentNode,endOffset);
+                    var allNodes = newRange.commonAncestorContainer.childNodes;
+                    for(nodeCount = 0;nodeCount<allNodes.length;nodeCount++)
+                    {
```
Comment (devashishmamgain) [inline webplugin/js/app/mck-sidebox-1.0.js]: @hardikparashar move { to above line and format the code

File: webplugin/js/app/mck-sidebox-1.0.js [hunk 10]
```diff
@@ -2480,19 +2480,135 @@ var CURRENT_GROUP_DATA={};
...
+                    var newRange = document.createRange();
+                    newRange.setStart(startNode,0);
+                    newRange.setEnd(currentNode,endOffset);
+                    var allNodes = newRange.commonAncestorContainer.childNodes;
+                    for(nodeCount = 0;nodeCount<allNodes.length;nodeCount++)
+                    { 
+                        if(allNodes[nodeCount].nodeType && allNodes[nodeCount].nodeType == "3" && allNodes[nodeCount].nodeValue == "") {
+                            emptyNodes++;
+                          }
+                      if(allNodes[nodeCount] == newRange.endContainer && allNodes[nodeCount].nodeType == "3") {
```
Comment (devashishmamgain) [inline webplugin/js/app/mck-sidebox-1.0.js]: @hardikparashar whats the purpose of allNodes[nodeCount].nodeType in the 'if' condition above this 'if'. if that can be undefined, same error can come here too.

File: webplugin/js/app/mck-sidebox-1.0.js [hunk 11]
```diff
@@ -2480,19 +2480,135 @@ var CURRENT_GROUP_DATA={};
...
+                                    normalSpan = document.getElementById('mck-normal-span');
+                                    normalSpan.innerHTML = preSpanContent;
+                                    warningSpan = document.getElementById('mck-text-warning-span');
+                                    warningSpan.innerHTML = spanContent;
+                                    selection = document.getSelection();
+                                    var targetNode;
+                                    if(caretNode == 0 && caretPosition > maxLength) {
+                                        targetNode = warningSpan.firstChild;
+                                        nodeOffset = caretPosition-maxLength;
+                                    }else{
```
Comment (devashishmamgain) [inline webplugin/js/app/mck-sidebox-1.0.js]: @hardikparashar code formatting required, please review all the new code lines.

File: webplugin/js/app/mck-sidebox-1.0.js [hunk 12]
```diff
@@ -2534,13 +2651,15 @@ var CURRENT_GROUP_DATA={};
                             }, 0);
                             return false;
                         }
-                    } else if (e.keyCode === 13) {
+                    } else if (e.keyCode === 13 && !CURRENT_GROUP_DATA.DISABLE_SEND_MESSAGE) {
                         e.preventDefault();
                         if (MCK_TYPING_STATUS === 1) {
                             mckInitializeChannel.sendTypingStatus(0, $mck_msg_inner.data('mck-id'));
                         }
                         ($mck_msg_sbmt.is(':disabled') && $mck_file_box.hasClass('vis')) ? alert('Please wait file is uploading.') : $mck_msg_form.submit();
```
Comment (devashishmamgain) [inline webplugin/js/app/mck-sidebox-1.0.js]: @hardikparashar move it to MCK_LABELS
Please wait file is uploading.

**Pull request merged**